docs/howto/how-to-make-tensorflow-plugin-with-ui/tensorflow-plugin-with-ui.md

```diff
@@ -0,0 +1,1488 @@
+
+```python
+# where bar = {"key_0": [0, 1]}
+try:
+  foo = bar[key_variable][2]
+except KeyError as e:
+  logger.error(f"An error occured: \n{repr(e)}")
+```
+
+This code will catch _an_ error; here, a `TypeError`. However, it will _not_ run our `except` we intend. Here, a couple more lines of code will go a long way for debugging.
+Your understanding of how the below items function would be best so you can efficiently debug your tool.
+If unsure, check out the [official Python docs](https://docs.python.org/3.8/tutorial/errors.html).
+
+```python
+except KeyError as e:
+  # Key error specific handling
+  logger.error(f"Key {EXPECTED_KEY} was not present: \n {repr(e)}")
+except Exception as e:
+  # Other exception handling
+  self.provider.io.warning(f"Unknown Error occured, see {LOG_PATH}/{PythonSDK.log} for details!")
+    logger.error(f"Key {EXPECTED_KEY} was not present: \n {repr(e)}")
+else:
+  # Code that runs when there are no exceptions
+  logger.error(f"Key {EXPECTED_KEY} was not present: \n {repr(e)}")
+finally:
+  # Code that will always run.
+  logger.error(f"Key {EXPECTED_KEY} was not present: \n {repr(e)}")
+```
+
+Whether you know already or know now, we recommend the above pattern (Note to reviewers: "when developing for designer" ?) with at least: 
+* `try/except` with a general case
+* A **log**\* statement logging the exception at the appropriate log level: `logger.INFO`, `logger.WARN`, etc
+* A `self.provider.io` **message**\* to notify the User an exception occurred and where to look
+  * NOTE: `self.provider.io.error(msg)` will immediately terminate the process! So ensure you log in and send any messages _before_ then or in your `finally` block if using one!
+
+\* Note the distinction here as well. A _log_ statement will automatically send your output to `PythonSdk.log`. A _message_ (read: `self.provider.io.[info | warn | error]`) will send your data output to Designer's IO.
+
+If you have multiple distinct sorts of errors or want to use your Exception derivatives, we recommend defining [decorators](https://pythonguide.readthedocs.io/en/latest/python/decorator.html) to make your exception handling **short, simple, and reusable across any tool you develop as a python module!**
+
+#### Adding a custom method: `setup_data()`
+
+Now we will add our first custom method to our plugin, `setup_data(self)`.
+It may be tempting to write your `DATA` mode code exactly where it will execute.
+**However** we strongly encourage following general coding best practices.
+Think, "How would I do this **in Python**?"
+Rather than, "How do I do this in the `ayx_python_sdk`?"
+When developing with the SDK, we encourage developers to use Python to its fullest potential.
+Therefore, we place our `DATA` mode code in its function, `setup_data(self)`.
+Separating your steps into functions naturally helps isolate problem areas in your code, makes debugging much easier, and (like mentioned above) makes even **more** of your code reusable! 
```

Comment:
TW review:

Now we add the first custom method to the plugin, `setup_data(self)`. It might be tempting to write your `DATA` mode code exactly where it will execute. However, we strongly encourage you to follow general coding best practices. Think, "How would I do this in Python?" Rather than, "How do I do this in the `ayx_python_sdk`?" When you develop with the SDK, we encourage you to use Python to its fullest potential. Therefore, place the `DATA` mode code in its function, `setup_data(self)`. When you separate your steps into functions, it naturally helps isolate problem areas in your code, makes debugging much easier, and makes even more of your code reusable!